### PR TITLE
v2 CLI - initial train and predict functionality

### PR DESCRIPTION
**Pull request by kevingreenman**
In progress. The following currently runs to completion without errors:

```
chemprop train -i tests/data/regression.csv
```

### CODE REVIEW

File: chemprop/v2/models/modules/readout.py [hunk 1]
```diff
@@ -104,6 +107,7 @@ def train_step(self, Z: Tensor) -> Tensor:
 class MveFFN(RegressionFFN):
     n_targets = 2
     _default_criterion = loss.MVELoss()
+    _default_metric = metrics.MSEMetric()
```
Comment (davidegraff) [inline chemprop/v2/models/modules/readout.py:110]: don't need to redefine because they all inherit from `RegressionFFN`

File: chemprop/v2/cli/utils/config.py [hunk 2]
```diff
@@ -2,5 +2,6 @@
 import os
 from pathlib import Path
 
+CHK_DIR = Path(os.getenv("CHEMPROP_CHK_DIR", "model_checkpoints"))
```
Comment (davidegraff) [inline chemprop/v2/cli/utils/config.py:5]:
```suggestion
CKPT_DIR = Path(os.getenv("CHEMPROP_CKPT_DIR", "model_checkpoints"))
```

File: chemprop/v2/cli/utils/config.py
Comment (davidegraff) [inline chemprop/v2/cli/utils/config.py:1]: can we rename this to `conf.py`. Seems to be a more common convention for package-level constants.

File: chemprop/v2/cli/utils/utils.py [hunk 3]
```diff
@@ -75,3 +75,14 @@ def validate_loss_function(
...
+
+def column_str_to_int(
+    columns: list,
+    header: list,
+    ) -> list:
+    if columns is None:
+        return None
+    if type(columns[0]) == str:
+        columns = [i for i,e in enumerate(header) if e in columns]
+    assert type(columns[0]) == int
```
Comment (davidegraff) [inline chemprop/v2/cli/utils/utils.py:87]:
```suggestion
    if all(isinstance(col, str) for col in columns):
        columns = [i for i, name in enumerate(header) if name in columns]
    if not all(isinstance(col, int) for col in columns):
        raise ValueError #TODO: add message
```
I'm assuming that `columns` comes in as one of  `None | list[str], list[int]` and if it's `list[str]` then it gets converted to `list[int]`?

File: chemprop/v2/cli/train.py [hunk 4]
```diff
@@ -156,7 +154,14 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
...
-        choices=['relu', 'leakyrelu', 'prelu', 'tanh', 'selu', 'elu'],  # TODO: should these be lowercase?
+        choices=[
+            "relu",
+            "leakyrelu",
+            "prelu",
+            "tanh",
+            "selu",
+            "elu",
+        ],  # TODO: should these be lowercase?
         help="activation function in message passing/FFN layers",
```
Comment (davidegraff) [inline chemprop/v2/cli/train.py:165]: activations are controlled by the `chemprop.v2.models.utils.Activation` enum

File: chemprop/v2/cli/train.py [hunk 5]
```diff
@@ -247,18 +255,20 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
...
+
     data_args.add_argument(
         "-t",
         "--dataset-type",
-        default="regression", 
-        action=RegistryAction(ReadoutRegistry),  # TODO: is this correct? The choices should be ['regression', 'classification', 'multiclass', 'spectra']
+        default="regression",
+        action=RegistryAction(
+            ReadoutRegistry
+        ),  # TODO: is this correct? The choices should be ['regression', 'classification', 'multiclass', 'spectra']
```
Comment (davidegraff) [inline chemprop/v2/cli/train.py:265]: I think we might want to change this over to `--task-type` because a `ReadoutFFN` controls the prediction task, which itself implicitly defines the dataset type. That is, there is no internal concept of a dataset type anymore

File: chemprop/v2/cli/train.py [hunk 6]
```diff
@@ -341,33 +355,29 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
...
-        "--loss-function",
-        action=RegistryAction(LossFunctionRegistry),
-    )
-    train_args.add_argument(
-        "--v-kl", 
+        "--v-kl",
         "--evidential-regularization",
-        type=float, 
+        type=float,
         default=0.2,  # TODO: the default in v1 was 0. Do we want it to default to 0.2 in v2?
```
Comment (davidegraff) [inline chemprop/v2/cli/train.py:363]: probably best to keep it the same to eliminate any surprises. In principle, however, it would be better to stick with the paper default

File: chemprop/v2/cli/train.py [hunk 7]
```diff
@@ -433,14 +423,17 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
         "--split",
         "--split-type",
         default="random",
-        choices=['random', 'scaffold'],  #'scaffold_balanced', 'predetermined', 'crossval', 'cv', 'cv-no-test', 'index_predetermined', 'random_with_repeated_smiles'],
+        choices=[
+            "random",
+            "scaffold",
+        ],  #'scaffold_balanced', 'predetermined', 'crossval', 'cv', 'cv-no-test', 'index_predetermined', 'random_with_repeated_smiles'],
```
Comment (davidegraff) [inline chemprop/v2/cli/train.py:429]:
```suggestion
        ],  #'scaffold_balanced', 'predetermined', 'crossval', 'cv', 'cv-no-test', 'index_predetermined', 'random_with_repeated_smiles'], # TODO: make data splitting CLI play nicely with astartes backend
```

File: chemprop/v2/cli/train.py [hunk 8]
```diff
@@ -433,14 +423,17 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
...
         help="Method of splitting the data into train/val/test.",
     )
     split_args.add_argument(
         "--split-sizes",
         type=list[float],
```
Comment (davidegraff) [inline chemprop/v2/cli/train.py:434]:
```suggestion
        type=float,
        nargs=3,
```

**Comment by hwpang** [comment]
With my changes, now `chemprop predict -i tests/data/regression_test_smiles.csv --checkpoint-path /path/to/checkpoint.ckpt` works.

**Comment by kevingreenman** [comment]
Fixes #410 with `**kwargs` in `MolGraphDataLoader`

**Comment by hwpang** [comment]
@kevingreenman Thanks for linking the issue! I actually added `**kwargs` to dataloader in my merged PR #471 (https://github.com/chemprop/chemprop/pull/471/commits/2cdf83cfe8b8be100e7e8f3734033a57fc6d15fd). I can rebase this branch with `v2/dev` if that's ok with you.

**Comment by kevingreenman** [comment]
@hwpang thanks for pointing that out, I had missed that detail in #471. I've closed that issue now. I have a couple minor changes I'm making to prediction now, then I can do the rebase and merge

File: chemprop/v2/cli/predict.py [hunk 9]
```diff
@@ -58,7 +59,7 @@ def add_predict_args(parser: ArgumentParser) -> ArgumentParser:
         "--output",
         "--preds-path",
         type=str,
```
Comment (davidegraff) [inline chemprop/v2/cli/predict.py:62]:
```suggestion
```

File: chemprop/v2/cli/predict.py [hunk 10]
```diff
@@ -153,32 +154,29 @@ def add_predict_args(parser: ArgumentParser) -> ArgumentParser:
...
     else:
         args.output = Path(args.output)
 
     return args
 
 
 def validate_predict_args(args):
-    pass
+    # TODO: onces args.checkpoint_dir and args.checkpoint are consolidated, need to change this as well. Not able to make this required in common.py as it may not be provided for training.
+    assert args.checkpoint_path is not None, "Must provide a checkpoint path for prediction."
```
Comment (davidegraff) [inline chemprop/v2/cli/predict.py:171]: replace with `raise ValueError`

File: chemprop/v2/cli/predict.py [hunk 11]
```diff
@@ -199,81 +197,61 @@ def main(args):
...
-
+    # TODO: add uncertainty and calibration
+    # if cal_dset is not None:
+    #     if args.dataset_type == "regression":
+    #         model.loc, model.scale = float(scaler.mean_), float(scaler.scale_)
+    #     predss_cal = trainer.predict(model, cal_loader)[0]
+
+    # TODO: might want to write a shared function for this as train.py might also want to do this.
+    df_test = pd.read_csv(args.test_path)
+    preds = np.vstack(predss)
```
Comment (davidegraff) [inline chemprop/v2/cli/predict.py:246]:
```suggestion
    preds = torch.concat(predss, 1).numpy()
```

File: chemprop/v2/cli/predict.py [hunk 12]
```diff
@@ -199,81 +197,61 @@ def main(args):
...
-    )
     logger.info(model)
 
     trainer = pl.Trainer(
         logger=None,
         enable_progress_bar=True,
         accelerator="auto",
         devices=args.n_gpu if torch.cuda.is_available() else 1,
-        max_epochs=args.epochs,
     )
```
Comment (davidegraff) [inline chemprop/v2/cli/predict.py:235]: wrap this whole block using [`torch.inference_mode()`](https://pytorch.org/docs/stable/generated/torch.inference_mode.html)

**Review by davidegraff — COMMENTED** [review]
just double checking (i should know since i wrote the code) that prediction of regression targets automatically applies the output transform (i.e., multiplies by `scale` and shifts by `loc`) and that binary classification outputs are passed to `nn.Sigmoid()`?

File: chemprop/v2/cli/predict.py (hunk 9 above)
Comment (kevingreenman) [reply chemprop/v2/cli/predict.py:62]: why?

**Comment by hwpang** [comment]
> just double checking (i should know since i wrote the code) that prediction of regression targets automatically applies the output transform (i.e., multiplies by `scale` and shifts by `loc`) and that binary classification outputs are passed to `nn.Sigmoid()`?

@davidegraff Thanks for double checking! For regression, the readout layer applies them in forward, see [here](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/v2/models/modules/readout.py#L97) for an example. For classification, the `Sigmoid` step is done in forward as well, see [here](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/v2/models/modules/readout.py#L162).

File: chemprop/v2/cli/train.py (hunk 6 above)
Comment (kevingreenman) [reply chemprop/v2/cli/train.py:363]: see #352 for explanation of why 0.0 was chosen as the default in v1 instead of 0.2 from the paper. I think we should stick with 0.0.

File: chemprop/v2/cli/train.py (hunk 4 above)
Comment (kevingreenman) [reply chemprop/v2/cli/train.py]: so this means all lowercase is correct, right? or does something need to change?

Comment (hwpang) [reply chemprop/v2/cli/train.py]: You can replace the choices with

```
list(Activation.keys())
```

to get all available ones.

File: chemprop/v2/cli/predict.py (hunk 12 above)
Comment (kevingreenman) [reply chemprop/v2/cli/predict.py:236]: @davidegraff is the right way to do this by putting the call to `trainer.predict` in a `with torch.inference_mode():` block (as [here](https://github.com/Lightning-AI/lightning/issues/11530#issuecomment-1067566245)), or to add a `inference_mode=False` argument to `pl.Trainer` (from the [documentation](https://lightning.ai/docs/pytorch/latest/common/trainer.html))?

I'm admittedly a bit confused by the documentation, since it says `Trainer(inference_mode=True)` is the default and `Trainer(inference_mode=False)` is equivalent to `torch.no_grad`, while I would expect the opposite to be true, unless I'm misunderstanding something...

**Comment by kevingreenman** [comment]
@davidegraff @hwpang I have addressed most of the comments and resolved merge conflicts. I confirmed the following commands run as expected:
```
chemprop train -i tests/data/regression.csv --epochs 1
chemprop predict -i tests/data/regression_test_smiles.csv -o test_preds.csv --checkpoint-path /path/to/checkpoint.ckpt
```

there are just 3 comments I'd like @davidegraff's feedback on before merging

File: chemprop/v2/cli/predict.py [hunk 13]
```diff
@@ -39,160 +40,156 @@ def add_args(cls, parser: ArgumentParser) -> ArgumentParser:
...
-        "-o", 
-        "--output-dir",
+        "-o",
+        "--output",
         "--preds-path",
-        dest="preds_path",
         type=str,
-        help="Path to CSV or PICKLE file where predictions will be saved.",
+        required=True,
+        help="Path to CSV or PICKLE file where predictions will be saved. If the file extension is .pkl, will be saved as a PICKLE file. If not provided and the test_path is /path/to/test/test.csv, predictions will be saved to /path/to/test/test_preds.csv.",
```
Comment (KnathanM) [inline chemprop/v2/cli/predict.py:64]: It doesn't make sense to require the user to give this argument and also tell them in the help that we have a default if they don't provide it. One of those should be removed.

File: chemprop/v2/cli/predict.py (hunk 9 above)
Comment (davidegraff) [reply chemprop/v2/cli/predict.py:62]: args are always parsed as strings, so this doesn't do anything

File: chemprop/v2/cli/predict.py (hunk 12 above)
Comment (davidegraff) [reply chemprop/v2/cli/predict.py:236]: yeah that's a good question. Looking at the source, it doesn't seem like the argument actually does _anything_. Even if it does do something in the future, it won't hurt us if our code is already inside a `with`-block (i.e., the PyTorch docs)

**Comment by davidegraff** [comment]
> @davidegraff @hwpang I have addressed most of the comments and resolved merge conflicts. I confirmed the following commands run as expected:
> there are just 3 comments I'd like @davidegraff's feedback on before merging

I only saw two comments? but otherwise they should be quick changes and then we're good to go.

File: chemprop/v2/cli/train.py (hunk 4 above)
Comment (davidegraff) [reply chemprop/v2/cli/train.py]: yeah just change this to
```python
choices=Activation.keys()
type=lambda x: x.lower()
```
The `type` arg is just a lambda that automatically puts all strings into lowercase so the input is case-insensitive.

File: chemprop/v2/cli/predict.py (hunk 12 above)
Comment (hwpang) [reply chemprop/v2/cli/predict.py:236]: @kevingreenman By reading the documentation, I think `Trainer(inference_mode=True)` means that `torch.inference_mode():` is used by Trainer internally during model evaluation (i.e., when `Trainer.validate`, `Trainer.test`, and `Trainer.predict` are called), while `Trainer(inference_mode=False)` uses `torch.no_grad():` during model evaluation. But I would like to know @davidegraff's thoughts as well.

File: chemprop/v2/cli/train.py [hunk 14]
```diff
@@ -799,16 +608,17 @@ def main(args):
         )
         criterion = readout_cls._default_criterion
 
+    print(readout_cls)
```
Comment (hwpang) [inline chemprop/v2/cli/train.py:611]: Can you remove this?
